### PR TITLE
DDS contoller frame and functions to control DDSs

### LINKED ISSUE #174: Frame with all DDS controllers: `DDSControllerFrame`
### Feature you want to implement
Using `DDSControllerWidget`, I will develop a frame with all DDS controllers.

### How the feature is implemented
`DDSControllerFrame` is a subclass of `QWidget`. It consists of several `DDSControllerWidget`s.

### Additional context
This issue is similar to #163.

See also: #152.

### LINKED ISSUE #175: Add functions in terms of DDS to `DeviceMonitorApp`
### Feature you want to implement
I will add functions to control `DDSControllerFrame` into `DeviceMonitorApp`, similar to #167.

### How the feature is implemented
The following elements will be added:
- A `DDSControllerFrame` to control DDS channels.
- A `DDSProfileThread` instance, which requests setting the frequency, amplitude, and phase to the proxy server.
- A `DDSAttenuationThread` instance, which requests setting the attenuation to the proxy server.
- A `DDSSwitchThread` instance, which requests turning on and off TTL switch, which controls the DDS output.

See also: #152

### PR DESCRIPTION
**Pull request by BECATRUE**
This closes #174, and closes #175.

I tested if controlling a DDS works well.

#### Screenshot
image

My laptop screen is so small that the frame couldn't be shown entirely, but in the bigger screen, it is shown well😅

### CODE REVIEW

**Comment by kangz12345** [comment]
😄 I think it is still too wide. How about move the switching checkbox below the spinboxes?
E.g., using a `QVBoxLayout`.
```
[freq / ampl / phase]
                         [ set ]
[     switching     ]
```

Or maybe we can just use the current version and change it if it is inconvenient!

File: iquip/apps/monitor.py [hunk 1]
```diff
@@ -619,25 +619,265 @@ def _setSwitchButtonText(self, on: bool):
...
+        # widgets
+        ddsWidgetLayout = QGridLayout()
+        for idx, (name, info) in enumerate(ddsInfo.items()):
+            ddsWidget = DDSControllerWidget(name, **info)
+            row, column = idx // numColumns, idx % numColumns
+            self.ddsWidgets[name] = ddsWidget
+            ddsWidgetLayout.addWidget(ddsWidget, row, column)
+        # layout
+        layout = QVBoxLayout(self)
+        layout.addLayout(ddsWidgetLayout)
```
Comment (kangz12345) [inline iquip/apps/monitor.py:663]: Why don't you just set the grid layout as the layout of `self`?

File: iquip/apps/monitor.py [hunk 2]
```diff
@@ -663,13 +907,24 @@ def __init__(
...
+            device, channel = map(info.get, ("device", "channel"))
+            self.ddsControllerFrame.ddsWidgets[name_].profileSet.connect(
+                functools.partial(self._setDDSProfile, device, channel)
+            )
+            self.ddsControllerFrame.ddsWidgets[name_].attenuationSet.connect(
+                functools.partial(self._setDDSAttenuation, device, channel)
+            )
+            self.ddsControllerFrame.ddsWidgets[name_].switchClicked.connect(
+                functools.partial(self._setDDSSwitch, device, channel)
+            )
```
Comment (kangz12345) [inline iquip/apps/monitor.py]: How about defining a local variable, e.g., `widget = self.ddsControllerFrame.ddsWidgets[name_]`?
I'm not sure if each statement can be written in a single line though 😄

File: iquip/apps/monitor.py [hunk 3]
```diff
@@ -689,13 +944,58 @@ def _setDACVoltage(self, device: str, channel: int, voltage: float):
...
-            device, channel, voltage: See _DACVoltageThread attributes section.
+            See _DACVoltageThread attributes section.
         """
         self.dacVoltageThread = _DACVoltageThread(
             device, channel, voltage, self.proxy_ip, self.proxy_port
         )
         self.dacVoltageThread.start()
 
-    def frames(self) -> Tuple[TTLControllerFrame, DACControllerFrame]:
+    @pyqtSlot(float, float, float, bool)
```
Comment (kangz12345) [inline iquip/apps/monitor.py]: Do you have any reference for this use?
This method itself does not match the signature `(float, float, float, bool)` but `functools.partial()` of it.

**Review by kangz12345 — CHANGES_REQUESTED** [review]

**Comment by BECATRUE** [comment]
> How about move the switching checkbox below the spinboxes?

Oh, it looks much better! I will come up with the best alignment.

File: iquip/apps/monitor.py (hunk 1 above)
Comment (BECATRUE) [reply iquip/apps/monitor.py:663]: I added the outermost layout for unity with `TTLControllerFrame` and `DACControllerFrame`.
Also, several widgets may be added later, such as description labels, so I want to remain it as it is.

File: iquip/apps/monitor.py (hunk 2 above)
Comment (BECATRUE) [reply iquip/apps/monitor.py]: Oh, it looks more legible! I will apply it.

Comment (BECATRUE) [reply iquip/apps/monitor.py:920]: Still, we cannot write them in a single line, but it became legible😃

File: iquip/apps/monitor.py (hunk 3 above)
Comment (BECATRUE) [reply iquip/apps/monitor.py]: Umm.. I think it should be changed into `(str, int, float, float, float, bool)`.
Thanks for reviewing it.

File: iquip/apps/monitor.py [hunk 4]
```diff
@@ -474,12 +474,12 @@ class DDSControllerWidget(QWidget):
         profileSet(frequency, amplitude, phase, switching):
           The default profile setting is set to frequency in Hz, amplitude, and phase.
           If switching is True, the current DDS profile is set to the default profile.
-        attenuationSet(attenuation): Current attenuation setting is set to attenuation.
+        attSet(attenuation): Current attenuation setting is set to attenuation.
```
Comment (kangz12345) [inline iquip/apps/monitor.py]: Why did you change `attenuation -> att`?
I don't think it is that long.
Also, this abbreviation is used inconsistently in this code (there are still many `attenuation`s)

**Review by kangz12345 — CHANGES_REQUESTED** [review]

File: iquip/apps/monitor.py (hunk 4 above)
Comment (kangz12345) [reply iquip/apps/monitor.py]: In contrast, you can use abbreviations for the GUI labels, e.g., "freq", "ampl", "att", etc.

File: iquip/apps/monitor.py [hunk 5]
```diff
@@ -522,19 +522,23 @@ def __init__(
...
         self.profileWidgets["switching"] = switchingCheckbox
-        profileLayout.addWidget(switchingCheckbox, alignment=Qt.AlignRight)
+        profileSetLayout.addWidget(switchingCheckbox)
         profileButton = QPushButton("Set", self)
-        profileLayout.addWidget(profileButton, alignment=Qt.AlignRight)
+        profileSetLayout.addWidget(profileButton, alignment=Qt.AlignRight)
+        profileLayout.addLayout(profileInfoLayout)
+        profileLayout.addLayout(profileSetLayout)
         # attenuation widgets
         attenuationBox = QGroupBox("attenuation", self)
```
Comment (kangz12345) [inline iquip/apps/monitor.py]: How about `"Attenuation"` (capitalized)?

**Review by kangz12345 — CHANGES_REQUESTED** [review]

File: iquip/apps/monitor.py (hunk 4 above)
Comment (BECATRUE) [reply iquip/apps/monitor.py]: I thought it is too long, but if you don't think so, I will revert them!

**Review by kangz12345 — APPROVED** [review]
LGTM!
